src/test/coinscachepair_tests.cpp

```diff
@@ -0,0 +1,193 @@
+    // Erase the nodes as we iterate through, but don't clear flags
+    // The flags will be cleared by the CCoinsCacheEntry's destructor
+    auto node{head.Next()};
+    for (auto expected{nodes.begin()}; expected != nodes.end(); expected = nodes.erase(expected)) {
+        BOOST_CHECK_EQUAL(&(*expected), node);
+        node = node->second.Next(/*clear_flags=*/false);
+    }
+    BOOST_CHECK_EQUAL(node, nullptr);
+
+    // Check that head's next is empty
+    BOOST_CHECK_EQUAL(head.Next(), nullptr);
+}
+
+BOOST_AUTO_TEST_CASE(linked_list_random_deletion)
+{
+    CCoinsCacheEntry head;
+    auto nodes{CreatePairs(head)};
+
+    // Create linked list head->n1->n2->n3->n4->nullptr
+    auto n1{nodes.begin()};
+    auto n2{std::next(n1)};
+    auto n3{std::next(n2)};
+    auto n4{std::next(n3)};
+
+    // Delete n2
+    // head->n1->n3->n4->nullptr
+    nodes.erase(n2);
+    // Check that n1 now points to n3, and n3 still points to n4
+    // Also check that flags were not altered
+    BOOST_CHECK_EQUAL(n1->second.GetFlags(), CCoinsCacheEntry::DIRTY);
+    BOOST_CHECK_EQUAL(n1->second.Next(), &(*n3));
+    BOOST_CHECK_EQUAL(n3->second.GetFlags(), CCoinsCacheEntry::DIRTY);
+    BOOST_CHECK_EQUAL(n3->second.Next(), &(*n4));
+
+    // Delete n1
+    // head->n3->n4->nullptr
+    nodes.erase(n1);
+    // Check that head now points to n3, and n3 still points to n4
+    // Also check that flags were not altered
+    BOOST_CHECK_EQUAL(n3->second.GetFlags(), CCoinsCacheEntry::DIRTY);
+    BOOST_CHECK_EQUAL(head.Next(), &(*n3));
+    BOOST_CHECK_EQUAL(n3->second.Next(), &(*n4));
+
+    // Delete n4
+    // head->n3->nullptr
+    nodes.erase(n4);
+    // Check that head still points to n3, and n3 points to nullptr
+    // Also check that flags were not altered
+    BOOST_CHECK_EQUAL(n3->second.GetFlags(), CCoinsCacheEntry::DIRTY);
+    BOOST_CHECK_EQUAL(head.Next(), &(*n3));
+    BOOST_CHECK_EQUAL(n3->second.Next(), nullptr);
+
+    // Delete n3
+    // head->nullptr
+    nodes.erase(n3);
+    // Check that head's next now points to nullptr
+    BOOST_CHECK_EQUAL(head.Next(), nullptr);
+}
+
+BOOST_AUTO_TEST_CASE(linked_list_add_flags)
+{
+    CCoinsCacheEntry head;
+    CoinsCachePair n1;
+    CoinsCachePair n2;
+
+    // Check that adding 0 flag has no effect
+    n1.second.AddFlags(0, n1, head);
+    BOOST_CHECK_EQUAL(n1.second.GetFlags(), 0);
+    BOOST_CHECK_EQUAL(n1.second.Next(), nullptr);
+    BOOST_CHECK_EQUAL(head.Next(), nullptr);
+
+    // Check that adding DIRTY flag inserts it into linked list and sets flags
+    n1.second.AddFlags(CCoinsCacheEntry::DIRTY, n1, head);
+    BOOST_CHECK_EQUAL(n1.second.GetFlags(), CCoinsCacheEntry::DIRTY);
+    BOOST_CHECK_EQUAL(n1.second.Next(), nullptr);
+    BOOST_CHECK_EQUAL(head.Next(), &n1);
+
+    // Check that adding FRESH flag on new node inserts it after head
+    n2.second.AddFlags(CCoinsCacheEntry::FRESH, n2, head);
+    BOOST_CHECK_EQUAL(n2.second.GetFlags(), CCoinsCacheEntry::FRESH);
```

Comment:
as mentioned before, after https://github.com/bitcoin/bitcoin/pull/28280/commits/cf5bfda650bd57636c7dd13f0ca03910d2aebe2e we can simplify these asserts
```C++
BOOST_CHECK(n2.second.IsFresh());
```

Reply:
I think the equivalent assert would be
```
BOOST_CHECK(n2.second.IsFresh() && !n2.second.IsDirty());
```

Reply:
the exact equivalent, yes, but you might not want to test the dirtiness part here - up to you, I don't mind either way.

Reply:
But since these are unit tests we are kind of testing the internals, so it would be good to catch if any other flags sneak in. With `IsFresh` it just checks that the FRESH flag is set, but any other flag could be set as well.